### PR TITLE
Sourmash gather implementation

### PR DESCRIPTION
**Pull request by elizabethmcd**
This pull request adds support for `sourmash gather` by requiring an input CSV listing the paths to the gather databases and corresponding lineage CSV files. I have not added full support for `taxannotate` just yet although it is added and commented out, because I need to find a test gather database that has some sort of matches to my test samples to ensure that that's working properly. 

It sets up the gather channels so that each sample-database pair runs separately, although `sourmash gather` can take a list of all the databases which could make the output easier to deal with. I couldn't figure out how to cleanly handle this as input while tracking what databases were ran, so the CSV input and splitting into different channels was the way I went. 

The test CI workflows should work as I put the sourmash k31 viral database in the Arcadia-Science/test-datasets repo as this was the smallest thing I could think to work with, the other option would be the contamination database we use in the seqqc workflow but I'm not sure if there's a lineage CSV for that? All TBD and would appreciate any and all feedback as I had to do quite a bit of trial and error to try to best figure out how to configure this.

### CODE REVIEW

**Review by taylorreiter — CHANGES_REQUESTED** [review]
I think this is going to end up being a very annoying review to receive...so i apologize in advance.

1. With the way sourmash gather runs, we really should only run it once with all desired databases in one go. There are two reasons for this. First, running it piecemeal will make it take longer in total with load times for the signature and other things. Second and maybe more importantly, it will change what results are given in some cases, and running with more databases is likely to lead to more accurate results (this is probably particularly true for Euk identification, where I have reason to suspect that a high portion of euk genomes in genbank, and therefore in the sourmash database, are contaminated).
2. as far as test data sets go, I there are probably better ones to use (as you pointed out). the seqqc contamination one does have a taxonomy file. Alternatively, you could use a scaled 100k cover database like you find here: https://osf.io/6zk3d/. These are much smaller than the full dbs and could be quick to search

File: modules/local/nf-core-modified/sourmash/taxannotate/main.nf
Comment (elizabethmcd) [reply modules/local/nf-core-modified/sourmash/taxannotate/main.nf:1]: addressed

File: modules/local/nf-core-modified/sourmash/gather/main.nf
Comment (elizabethmcd) [reply modules/local/nf-core-modified/sourmash/gather/main.nf]: no longer relevant I believe

**Comment by elizabethmcd** [comment]
I have fixed the channel input for `sourmash gather` and `sourmash taxonomy annotate` so that all databases that are input to a CSV are run at the same time for each sample. The first part is addressed by fixing the test-datasets repo to include sourmash databases that are small and do hit against the samples (in this case they hit against the reads but not the assemblies in some cases, so I've also changed to control for that), in this PR: https://github.com/Arcadia-Science/test-datasets/pull/21. The second part is addressed by reading in the CSV and using the `.collect()` and `.toList()` methods.

File: modules/local/check_sourmash_dbs.nf
Comment (elizabethmcd) [reply modules/local/check_sourmash_dbs.nf:1]: this is no longer relevant and has been removed

File: modules/local/nf-core-modified/sourmash/gather/main.nf
Comment (elizabethmcd) [reply modules/local/nf-core-modified/sourmash/gather/main.nf:3]: I don't think I understand this comment - when you re-review my updated PR can you see if this is still an issue and clarify?

File: subworkflows/local/sourmash_profiling.nf
Comment (mertcelebi) [inline subworkflows/local/sourmash_profiling.nf:9]:
```suggestion
    seqtype         // "reads" or "assembly"
```

File: subworkflows/local/sourmash_profiling.nf
Comment (mertcelebi) [inline subworkflows/local/sourmash_profiling.nf:10]:
```suggestion
    databases_csv  // path(sourmash_dbs_csv) CSV file of database_path,lineage_path
```

File: subworkflows/local/sourmash_profiling.nf
Comment (mertcelebi) [inline subworkflows/local/sourmash_profiling.nf:21]: I don't think you need both collect and toList. Just toList may be sufficient.

File: subworkflows/local/sourmash_profiling.nf
Comment (mertcelebi) [inline subworkflows/local/sourmash_profiling.nf:54]: Is there a reason to do this and not provide the list of paths as another separate variable? Perhaps I don't understand the final output shape of `ch_input_gather`, could you add a comment about it?

Also in the underlying `SOURMASH_GATHER`, you seem to refer to `databases` as a single database path. Is that the case? Or is it a list of paths and if so does it stage all of them correctly?

File: subworkflows/local/sourmash_profiling.nf
Comment (mertcelebi) [inline subworkflows/local/sourmash_profiling.nf:68]: same comment as above

File: subworkflows/local/sourmash_profiling.nf
Comment (mertcelebi) [inline subworkflows/local/sourmash_profiling.nf:75]:
```suggestion
    // TODO: calls a script that outputs an HTML document for Rmarkdown rendering???
```

**Review by mertcelebi — CHANGES_REQUESTED** [review]
Just left a bunch of questions, I am a lil bit confused about how all of this combines and collects work.

File: subworkflows/local/sourmash_profiling.nf
Comment (elizabethmcd) [reply subworkflows/local/sourmash_profiling.nf:21]: In my testing if you don't collect before setting as a list it doesn't put them together as one so that when you do combine it won't keep the entire list of databases together and gets confused how to input it all

File: subworkflows/local/sourmash_profiling.nf
Comment (elizabethmcd) [reply subworkflows/local/sourmash_profiling.nf:54]: It is a list of paths, so when you give it multiple database paths in the CSV it will run them all together once per sample, and currently how I have it set up it stages it correctly yes. 
So if you have 2 databases and 2 samples, there should only be 2 ch_input_gather feeds, one for each sample so that it runs all the databases once for each sample.

File: subworkflows/local/sourmash_profiling.nf
Comment (elizabethmcd) [reply subworkflows/local/sourmash_profiling.nf:68]: same logic so that all the databases/taxonomy options that are input in the sourmash_dbs CSV are run once per sample, which is what combine is doing once the `.collect()` and `.toList()` for the databases or lineage CSVs is set up and it's queuing one sample per list of databases/taxonomy CSVs.

File: modules/local/nf-core-modified/sourmash/gather/main.nf
Comment (taylorreiter) [reply modules/local/nf-core-modified/sourmash/gather/main.nf:3]: `process_low` won't have enough RAM if you run sourmash with the 5 genbank databases

if you keep this as-is, you'll need to have logic to increase the ram by multiplying by number of failures/retries or whatever

File: modules/local/nf-core-modified/sourmash/taxannotate/main.nf
Comment (taylorreiter) [inline modules/local/nf-core-modified/sourmash/taxannotate/main.nf:4]: thank you this is great!

File: subworkflows/local/sourmash_profiling.nf
Comment (taylorreiter) [inline subworkflows/local/sourmash_profiling.nf:10]: maybe add to the comment that the CSV can contain multiple databases?

**Review by taylorreiter — COMMENTED** [review]
This LGTM now, all of my comments are addressed! I'm just comment bc I'll let @mertcelebi approve it after he re-reviews

File: subworkflows/local/sourmash_profiling.nf
Comment (elizabethmcd) [reply subworkflows/local/sourmash_profiling.nf:21]: If you run `.toList()` by itself the resulting input gather channel looks like: 
`[[id:comm_1_T1, single_end:false], /Users/emcdaniel/Desktop/Arcadia/workflows/metagenomics/work/b6/4cf76eede4b448f83f46fc2892f7db/comm_1_T1.reads.sig, /Users/emcdaniel/Desktop/Arcadia/workflows/test-datasets/metagenomics/sourmash_dbs/GCF_001457635.1.db.zip, /Users/emcdaniel/Desktop/Arcadia/workflows/test-datasets/metagenomics/sourmash_dbs/GCF_003697165.2.db.zip]`

whereas with both `.collect()` and `.toList()` it looks like
`[[id:comm_1_T1, single_end:false], /Users/emcdaniel/Desktop/Arcadia/workflows/metagenomics/work/b6/4cf76eede4b448f83f46fc2892f7db/comm_1_T1.reads.sig, [/Users/emcdaniel/Desktop/Arcadia/workflows/test-datasets/metagenomics/sourmash_dbs/GCF_001457635.1.db.zip, /Users/emcdaniel/Desktop/Arcadia/workflows/test-datasets/metagenomics/sourmash_dbs/GCF_003697165.2.db.zip]]
` where I need the brackets around the list of databases, and for whatever reason just `.toList()` won't get me there. Same with just `.collect()` by itself. Then it works properly running all databases once for each input signature (once per sample)

File: subworkflows/local/sourmash_profiling.nf
Comment (elizabethmcd) [reply subworkflows/local/sourmash_profiling.nf:54]: and in sourmash gather it says `path (databases)` but it gets passed as a list and it stages correctly, I went into the work directory for this process and checked them

**Review by mertcelebi — APPROVED** [review]

**Pull request merged**
